### PR TITLE
Fixed #24168 -- Allowed selecting a template engine in a few APIs.

### PR DESCRIPTION
**Pull request by aaugustin**
Specifically in rendering shortcuts, template responses, and class-based
views that return template responses.

### CODE REVIEW

**Comment by aaugustin** [comment]
I'm wondering where the tests should live.

There isn't a test application dedicated to testing multiple template engines. I don't want to add the boilerplate for setting up multiple template engines in the shortcuts, template responses and class-based views tests either.

File: docs/ref/class-based-views/mixins-simple.txt [hunk 1]
```diff
@@ -49,6 +49,17 @@ TemplateResponseMixin
         a ``template_name`` will raise a
         :class:`django.core.exceptions.ImproperlyConfigured` exception.
 
+    .. attribute:: template_engine
+
+        The :setting:`NAME <TEMPLATES-NAME>` of a template engine to use for
+        loading the template. ``template_engine`` is passed as the ``using``
+        keyword argument to ``response_class``. Default is ``None``, which
+        tells Django to search the template with all configured engines.
```
Comment (timgraham) [inline docs/ref/class-based-views/mixins-simple.txt]: "for the template in all"?

File: docs/ref/class-based-views/mixins-simple.txt [hunk 2]
```diff
@@ -49,6 +49,17 @@ TemplateResponseMixin
...
+
+        .. versionadded:: 1.8
```
Comment (timgraham) [inline docs/ref/class-based-views/mixins-simple.txt]: Usually we put this after `.. attribute::` without any content.

File: docs/ref/template-response.txt [hunk 3]
```diff
@@ -102,9 +102,17 @@ Methods
         be extracted from ``content_type``, and if that is unsuccessful, the
         :setting:`DEFAULT_CHARSET` setting will be used.
 
-    .. versionadded:: 1.8
```
Comment (timgraham) [inline docs/ref/template-response.txt]: I think I'd keep this at the current indent level and just have it at the bottom and mention both parameters, e.g. "The charset and using parameters were added."

File: docs/ref/template-response.txt [hunk 4]
```diff
@@ -235,9 +243,17 @@ Methods
...
```
Comment (timgraham) [inline docs/ref/template-response.txt]: as above

File: docs/topics/http/shortcuts.txt [hunk 5]
```diff
@@ -163,6 +171,14 @@ Optional arguments
 ``status``
     The status code for the response. Defaults to ``200``.
 
+``using``
+    The :setting:`NAME <TEMPLATES-NAME>` of a template engine to use for
+    loading the template.
+
+    .. versionadded:: 1.8
+
+        The ``using`` parameter was added.
```
Comment (timgraham) [inline docs/topics/http/shortcuts.txt]: combine with the "status" note below?

File: docs/topics/http/shortcuts.txt [hunk 6]
```diff
@@ -77,6 +77,14 @@ Optional arguments
        The ``current_app`` argument is deprecated. Instead you should set
        ``request.current_app``.
 
+``using``
+    The :setting:`NAME <TEMPLATES-NAME>` of a template engine to use for
+    loading the template.
+
+    .. versionadded:: 1.8
```
Comment (timgraham) [inline docs/topics/http/shortcuts.txt]: inconsistent with dirs note

**Comment by timgraham** [comment]
Re: tests, I have thought that a place to put private testing utilities would be useful (currently we've been sticking them in `django.test.utils`. I think it would be better to share the boilerplate and put the tests with the testing of the rest of the function rather than testing the multiple templates engine aspect of these features in one app.

File: django/shortcuts.py [hunk 7]
```diff
@@ -80,7 +83,8 @@ def render(request, template_name, context=None,
                 context_instance._current_app = current_app
 
         content = loader.render_to_string(
-            template_name, context, context_instance, dirs, dictionary)
+            template_name, context, context_instance, dirs, dictionary
```
Comment (zachborboa) [inline django/shortcuts.py]: Missing trailing comma after `dictionary`.

**Comment by aaugustin** [comment]
This should be good to go now. Open questions:

1. Does the definition of `require_jinja2` look reasonable? It doesn't do much, but I don't need it to do anything else either.
2. Is `template_engine` a good name for the attribute of `TemplateResponseMixin`? `using` sounded too generic to me.

I can ask these questions on django-developers if more input seems needed.

**Comment by carljm** [comment]
I did a quick skim, not a detailed review, but with particular attention to the two issues you called out, and I think both are good as-is.

File: docs/ref/template-response.txt [hunk 8]
```diff
@@ -102,9 +102,13 @@ Methods
...
+    ``using``
+        The :setting:`NAME <TEMPLATES-NAME>` of a template engine to use for
+        loading the template.
 
-        The ``charset`` parameter was added.
+    .. versionchanged:: 1.8
+
+        The ``charset`` and ``using`` parameter were added.
```
Comment (timgraham) [inline docs/ref/template-response.txt]: parameters

File: docs/ref/template-response.txt [hunk 9]
```diff
@@ -235,9 +239,13 @@ Methods
...
 
-    .. versionadded:: 1.8
+    ``using``
+        The :setting:`NAME <TEMPLATES-NAME>` of a template engine to use for
+        loading the template.
+
+    .. versionchanged:: 1.8
 
-        The ``charset`` parameter was added.
+        The ``charset`` and ``using`` parameter were added.
```
Comment (timgraham) [inline docs/ref/template-response.txt]: parameters

File: docs/topics/http/shortcuts.txt [hunk 10]
```diff
@@ -159,9 +167,13 @@ Optional arguments
...
     The status code for the response. Defaults to ``200``.
 
+``using``
+    The :setting:`NAME <TEMPLATES-NAME>` of a template engine to use for
+    loading the template.
+
 .. versionchanged:: 1.8
 
-   The ``status`` parameter was added.
+   The ``status`` and ``using`` parameters were added.
```
Comment (timgraham) [inline docs/topics/http/shortcuts.txt]: 3rd time a charm!

**Comment by timgraham** [comment]
At first I was going to suggest renaming `require_jinja2` to `requires_jinja2`, but since it does more than just check if it's installed, I think the name is okay. The alternative that comes to mind would be `setup_jinja2`, but I don't really care for that over what you've picked.

`template_engine` also seems clear to me.

File: django/test/utils.py [hunk 11]
```diff
@@ -3,11 +3,16 @@
 import re
 import sys
 import time
-from unittest import skipUnless
+from unittest import skipIf, skipUnless
 import warnings
 from functools import wraps
 from xml.dom.minidom import parseString, Node
 
+try:
```
Comment (timgraham) [inline django/test/utils.py]: The new import order guidelines haven't merged yet, but they call for putting try/except imports last (as isort can't parse them to order them elsewhere).

**Pull request merged**